x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/elastic/action/ModelStrategyFactory.java

```diff
@@ -0,0 +1,187 @@
+import org.elasticsearch.xpack.inference.services.elastic.request.ElasticInferenceServiceRerankRequest;
+import org.elasticsearch.xpack.inference.services.elastic.rerank.ElasticInferenceServiceRerankModel;
+import org.elasticsearch.xpack.inference.services.elastic.sparseembeddings.ElasticInferenceServiceSparseEmbeddingsModel;
+import org.elasticsearch.xpack.inference.telemetry.TraceContext;
+
+import static org.elasticsearch.xpack.inference.services.elastic.ElasticInferenceService.ELASTIC_INFERENCE_SERVICE_IDENTIFIER;
+import static org.elasticsearch.xpack.inference.services.elastic.request.ElasticInferenceServiceRequest.extractRequestMetadataFromThreadContext;
+
+record ModelStrategyFactory(ServiceComponents serviceComponents) {
+
+    public interface Strategy<T extends ElasticInferenceServiceModel> {
+        RequestManager createRequestManager(
+            T model,
+            ServiceComponents serviceComponents,
+            TraceContext traceContext,
+            CCMAuthenticationApplierFactory.AuthApplier authApplier
+        );
+
+        String requestDescription();
+    }
+
+    private static final String SPARSE_EMBEDDINGS_REQUEST_DESCRIPTION = Strings.format(
+        "%s sparse embeddings",
+        ELASTIC_INFERENCE_SERVICE_IDENTIFIER
+    );
+
+    private static final Strategy<ElasticInferenceServiceSparseEmbeddingsModel> SPARSE_EMBEDDINGS_STRATEGY = new Strategy<>() {
+        @Override
+        public RequestManager createRequestManager(
+            ElasticInferenceServiceSparseEmbeddingsModel model,
+            ServiceComponents serviceComponents,
+            TraceContext traceContext,
+            CCMAuthenticationApplierFactory.AuthApplier authApplier
+        ) {
+            return new ElasticInferenceServiceSparseEmbeddingsRequestManager(model, serviceComponents, traceContext, authApplier);
+        }
+
+        @Override
+        public String requestDescription() {
+            return SPARSE_EMBEDDINGS_REQUEST_DESCRIPTION;
+        }
+    };
+
+    private static final String RERANK_REQUEST_DESCRIPTION = Strings.format("%s rerank", ELASTIC_INFERENCE_SERVICE_IDENTIFIER);
+
+    private static final ResponseHandler RERANK_HANDLER = new ElasticInferenceServiceResponseHandler(
+        RERANK_REQUEST_DESCRIPTION,
+        (request, response) -> ElasticInferenceServiceRerankResponseEntity.fromResponse(response)
+    );
+
+    private static final Strategy<ElasticInferenceServiceRerankModel> RERANK_STRATEGY = new Strategy<>() {
+        @Override
+        public RequestManager createRequestManager(
+            ElasticInferenceServiceRerankModel model,
+            ServiceComponents serviceComponents,
+            TraceContext traceContext,
+            CCMAuthenticationApplierFactory.AuthApplier authApplier
+        ) {
+            var metadata = extractRequestMetadataFromThreadContext(serviceComponents.threadPool().getThreadContext());
```

Comment:
I think there might have been a subtle lose of context issue here. Prior to the refactor this extract call was done in the lambda which is being executed on a separate thread. I think moving this out of the lambda will ensure we're using the right thread context of the original request.